src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -225,26 +219,70 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
+            {
+                JsonPropertyInfo other = cache[jsonPropertyInfo.NameAsString];
+
+                if (other.IsIgnored)
+                {
+                    // Overwrite previously cached property since it has [JsonIgnore].
+                    cache[jsonPropertyInfo.NameAsString] = jsonPropertyInfo;
+                }
+                else if (
+                    // Does the current property have `JsonIgnoreAttribute`?
+                    !jsonPropertyInfo.IsIgnored &&
+                    // Is the current property hidden by the previously cached property
+                    // (with `new` keyword, or by overriding)?
+                    other.MemberInfo!.Name != memberName &&
+                    // Was a property with the same CLR name was ignored? That property hid the current property,
+                    // thus, if it was ignored, the current property should be ignored too.
+                    ignoredProperties?.Contains(memberName) != true)
+                {
+                    // We throw if we have two public properties that have the same JSON property name, and neither have been ignored.
+                    ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameConflict(Type, jsonPropertyInfo);
+                }
+                // Ignore the current property.
+            }
+
+            if (jsonPropertyInfo.IsIgnored)
+            {
+                (ignoredProperties ??= new HashSet<string>()).Add(memberName);
+            }
+        }
+
         private void InitializeConstructorParameters(Dictionary<string, JsonPropertyInfo> propertyCache, ConstructorInfo constructorInfo)
         {
             ParameterInfo[] parameters = constructorInfo!.GetParameters();
             Dictionary<string, JsonParameterInfo> parameterCache = CreateParameterCache(parameters.Length, Options);
 
             foreach (ParameterInfo parameterInfo in parameters)
             {
-                PropertyInfo? firstMatch = null;
+                MemberInfo? firstMatch = null;
                 bool isBound = false;
 
                 foreach (JsonPropertyInfo jsonPropertyInfo in PropertyCacheArray!)
                 {
                     // This is not null because it is an actual
                     // property on a type, not a "policy property".
-                    PropertyInfo propertyInfo = jsonPropertyInfo.PropertyInfo!;
+                    MemberInfo memberInfo = jsonPropertyInfo.MemberInfo!;
 
-                    string camelCasePropName = JsonNamingPolicy.CamelCase.ConvertName(propertyInfo.Name);
+                    string camelCasePropName = JsonNamingPolicy.CamelCase.ConvertName(memberInfo.Name);
 
                     if (parameterInfo.Name == camelCasePropName &&
-                        parameterInfo.ParameterType == propertyInfo.PropertyType)
+                        parameterInfo.ParameterType == jsonPropertyInfo.DeclaredPropertyType)
```

Comment:
@layomia this line caused failures because `CreateIgnoredPropertyPlaceholder` uses `byte` for placeholder properties. The easiest way to fix it is to do `switch` on `memberInfo` and get its type, but I don't like it.

Reply:
Why don't you like it? I think switching on `memberInfo` is fine, since we previously got it from `propertyInfo`, not `jsonPropertyInfo` (perhaps for the same reason).

Reply:
Not obvious that an ignored JSON property doesn't represent the corresponding CLR property, plus code bloating because we should repeat type determination when there should be one entry point.

Reply:
We can revisit this.